File: src/jsp/domain/solution/schedule.py
from dataclasses import dataclass
from jsp.domain.solution.schedule_item import ScheduleItem


@dataclass(slots=True, frozen=True)
class Schedule:
    makespan: int
    items: tuple[ScheduleItem, ...]

    def __str__(self) -> str:
        header = (
            f"Optimal Makespan: {self.makespan}\n\n"
            f"{'Job':<6}{'Op':<6}{'Machine':<10}{'Start':<8}{'End':<8}{'Duration':<10}\n"
            + "-" * 50
        )
        sorted_items = sorted(self.items, key=lambda x: (x.job, x.op))
        rows = "\n".join(str(item) for item in sorted_items)
        return f"{header}\n{rows}"

File: src/domain/solution/schedule.rs
use std::fmt::Display;

use super::ScheduleItem;

/// Representa la solución o programación completa obtenida para una instancia JSP.
///
/// Almacena la lista de operaciones programadas ([`ScheduleItem`]) y el makespan global
/// resultante (tiempo total para completar todos los trabajos).
pub struct Schedule {
    makespan: usize,
    items: Vec<ScheduleItem>,
}

// Inicializadores
impl Schedule {
    pub fn new(makespan: usize, items: Vec<ScheduleItem>) -> Self {
        Self { makespan, items }
    }
}

impl Display for Schedule {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut sorted_items = self.items.clone();
        sorted_items.sort_by_key(|item| (item.job(), item.op()));

        writeln!(f, "Optimal Makespan: {}\n", self.makespan)?;
        writeln!(
            f,
            "{:<6}{:<6}{:<10}{:<8}{:<8}{:<10}",
            "Job", "Op", "Machine", "Start", "End", "Duration"
        )?;
        writeln!(f, "{}", "-".repeat(50))?;

        for item in sorted_items {
            writeln!(f, "{}", item)?;
        }

        Ok(())
    }
}
